Add micro qr code detector

corbers opened the pull request:
This PR adds a micro QR code detector as the first step towards adding micro QR code reading support.  The PR contains unit tests using artificial codes to check that the basic functionality of each class works as expected.  Tests using real world bar codes will be added later when the decoder is complete.

Issues that I'm aware of...

1. I've used a sub-directory of decoder within the microqrcode directory when the nu-book approach appears to be to put detector, decoder and encoder in a single directory.
2. Each filename should begin with a shortened version of the code, e.g. MQR.

I'm sure there are other things to discuss!

axxel commented:
Thanks. And here we go with some high level stuff first:
 * your 3 latest commits in this row don't fit into the "rewriting history" strategy that I mentioned earlier and I'm fine with just adding more changes like those but then squash them before merging as I see no value in preserving a history of e.g. code reformatting or comment fixing. For a clean history those two would have be merged into the earlier commits.
 * your 2 mentioned issues with the file layout (sub directory and file name pattern) are indeed something I'd like to see harmonized with the rest of the code base, so why not start here.
 * We don't use exceptions to report detection failures anywhere in the library. That was @huycn original decision when he ported the code from Java and I came to completely agree with it. Exceptions should not be used to manage 'non-exceptional' control flow (e.g. for performance reasons). And not finding a specific symbol in an image is in fact the most likely scenario with any given input ;). Please use the `DetectorResult` default constructor and the `isValid()` member to manage that control flow.
 * We use `#pragma once` instead of include guard macros
 * Please consider the `ResultPoint` class to be deprecated and only still around for the old PDF417 code and use the underlying `PointF` (or `PointI` if appropriate) instead. There are also convenience functions to calculate distance, etc. in `Point.h`. If you deem this to be too invasive and time consuming a change, I'll take care of that later.
 * Style question: We started to drop braces on one-line control blocks (like `if` statements) a while back. The current code base is not consistent in this regard but for new code it would be a nice to have (again: no merge blocker for me)
 * The `FinderPatternFinder.cpp` code can pretty much go away as it is a copy of some 'old' code that I replaced a while back and it does exactly what the QRCode code is doing already. Not only could the code be shared between the two implementations but actually the processing results as this finder pattern scanning is an expansive (if not the dominant) part of the detection. You can have a look at `QRDetector.cpp:FindFinderPatterns()`. I would in fact rather merge a plain copy of that function and later deduplicate it than reintroduce the `FinderPatternFinder` code. What is possible here pre-merging obviously depends a lot on how much time you want to spend -> your call.
 * making copies of the `BitMatrix` should definitively be avoided. The other code uses a more "functional" style of passing a reference to that data than the 'object oriented' approach of your (and the upstream Java) code.
 
If you are one with whatever of the above you want to spend time one, I'll have another look at more fine grained level.

axxel commented:
> Thanks for the feedback. I'll work on these tomorrow.
Perfect.

> Regarding your first point - do you want me to squash these commits together now or is this something you'll do during the merge?

There is a "squash and merge" button on github, very easy operation, so nothing to worry about at the moment.

> For the rest of your comments I'll probably use a commit per comment if that's ok?

Sure.

corbers commented:
I've responded to most of the comments you made but there are two that I've left for now.
- Refactor `FinderPatternFinder.cpp` to use `QRDetector.cpp:FindFinderPatterns()`
- Replace `ResultPoint` with `PointF` or `PointI`

The pattern finder refactor I'd prefer to do once the decoder is in place because then I can test the changes on real barcodes.  At the moment I'm using artificial codes generated in the unit test code.
The ResultPoint change I would also prefer to do once the decoder is in place and I can run more test data through.  This change should also be simpler once the refactor of the pattern finder has been carried out.
Thoughts?

axxel reviewed (changes requested):
Your argument to exchange functionality (`ResultPoint` and the `FinderPatternFinder`) only after a full test-cycle is possible, is indeed valid.

Depending on what will be left after those changes are made, I suggest to also use the `Quadrilateral` abstraction instead of `vector<Point>` later on where appropriate.

corbers commented on core/src/microqrcode/MQRCornerFinder.cpp:
```diff
@@ -0,0 +1,273 @@
+ *
+ * @author Christian Braun
+ */
+CornerFinder::CornerFinder() {}
```
Agreed.  I think the usage of `using ClassName;` was inherited from the original Java to C++ port that used some sort of tool to automate the process.  I've updated all the new files I added to the project to use `namespace ZXing::MicroQRCode`.

corbers commented on core/src/microqrcode/MQRCornerFinder.cpp:
```diff
@@ -0,0 +1,273 @@
+		return std::vector<ResultPoint>{};
+
+	ResultPoint centerEnclosingRectA, centerEnclosingRectB, centerEnclosingRectC, centerEnclosingRectD;
+	if (!DetectWhiteRect(image, std::lround(center.getEstimatedModuleSize() * 4), std::lround(center.x()),
```
We have 43 samples that have various amounts of rotation.  Samples rotated up to 30 degrees are OK.  Samples with more rotation can often fail.  It seems that a rotation of 45 + n*90 (where n is between 0 and 3) is particularly challenging.  I hope that using some of the techniques you have for QR codes will help improve things but I predict there will be more work required to get the performance on rotated micro QR codes to a good level.

corbers commented on core/src/microqrcode/MQRCornerFinder.h:
```diff
@@ -0,0 +1,54 @@
+
+namespace MicroQRCode {
+
+class CornerFinder
```
Agreed and done.

corbers commented on core/src/microqrcode/MQRDetector.cpp:
```diff
@@ -0,0 +1,249 @@
+#include "MQRFinderPatternFinder.h"
+#include "PerspectiveTransform.h"
+
+#include <algorithm> // vs12, std::min and std:max
```
It either means Visual Studio 2012 or Visual Studio version number 12.  I'm guessing that the algorithm header file needed to be included to get the software building with this compiler but for GCC it didn't.  Anyway, I've removed the comment and simplified the C++ header files included.

corbers commented on core/src/microqrcode/MQRCornerFinder.h:
```diff
@@ -0,0 +1,54 @@
+	std::vector<ResultPoint> find(const BitMatrix& image, const FinderPattern& center) const;
+
+private:
+	ResultPoint calculateDirection(const BitMatrix& image, const FinderPattern& center) const;
```
Done.

axxel commented:
I've been offline over the weekend, hence the silence. Thanks for the work. Looks good to me. I'll not merge it right away, since I'm about to make a 1.3 release first. But I believe there is no reason not to start on the next PR including the decoder part. I have no idea but I could imagine that the decoding of QR and microQR are (almost) the same? If that should be the case, I'd argue it makes sense to see what existing functionality could be used as is, without introducing a copy thereof first that we'd remove later anyway.

axxel commented:
Thinking about it: I believe it would actually make more sense to add the decoder part to this PR as well. Then we could maybe even replace the FinderPatternFinder before merging it in. And it would guarantee that something actually extending the functionality of the library ends up in the repo. What do you think?

corbers commented:
That's fine with me.  In the past few days I've been adding the decoder functionality to a separate branch in anticipation of another pull request.  I'll move these changes over to this pull request today.  Hopefully I can push something later today.

Annoyingly, there are enough small differences between QR and micro QR codes that a lot of the QR classes/functions have a micro equivalent.  I'm sure this can be improved but it will require changes to the QR files as well.  This raises the question of if code is shared between QR and micro QR then where is the common code stored?  Should the micro QR files depend on the QR files or should the common code get pushed up a directory to the source root?  Or maybe all the QR and micro QR files should reside in the qrcode directory?

axxel commented:
> Or maybe all the QR and micro QR files should reside in the qrcode directory?

That actually makes more and more sense to me. Given the discussed finder pattern finder situation, especially the fact that we can use part of the processing for both variants, suggests that to me. I'd still consider those two as separate `BarcodeFormats` and would also propose to keep the `MicroQRCode` namespace as well as the `MQR` file prefix, but lowering the bar for sharing common code seems very natural to me now. Sorry for coming only now to this conclusion.

Still: if the differences are more than trivial, we can surely put a copy of parts of the decoder in and cleanup later. I just thought there might be functions that are simply doing exactly the same thing already.

corbers commented:
There are a few updates in this push...
- Moved all files to qrcode directory.
- Added decoder.
- Separated ErrorCorrectionLevel and FormatInformation classes into definition and construction files so that QR and micro QR versions can be constructed separately but treated the same in other parts of the implementation.  Potentially this approach could be used for Version as well which should remove the need for two GetDataBlock functions and associated files ([QR|MQR]DataBlock.[cpp.h]).
- Added micro QR samples and hooked them into ReaderTest.

My next step is to look at the finder pattern finder improvements you suggested.  If you have any comments on the recent commits then that would be great!

axxel commented on core/src/qrcode/MQRDecoder.cpp:
```diff
@@ -0,0 +1,392 @@
+ *
+ * <p>See ISO 18004:2006, 6.4.3 - 6.4.7</p>
+ */
+ZXING_EXPORT_TEST_ONLY DecoderResult DecodeBitStream(ByteArray&& bytes, const Version& version,
```
This file is almost identical to `QRDecoder.cpp`. The few differences in this function should easily be manageable with a boolean flag discriminating between standard and micro version. This would also require to use the `QRCode::CodecMode` which should be possible (I think). I might be overlooking something and would need to actually get this code into my IDE and start merging it to see any issues.

axxel commented on core/src/qrcode/MQRDecoder.cpp:
```diff
@@ -0,0 +1,392 @@
+	return DecoderResult(std::move(bytes), std::move(result)).setEcLevel(ToString(ecLevel));
+}
+
+static DecoderResult DoDecode(const BitMatrix& bits, const Version& version, const std::string& hintedCharset,
```
Also the same code except for the different implementation of the `ReadFormatInformation` and other helper functions. I think a bool flag either passed to those or used to descriminate between calling them from either of the two namespaces should be the simplest way to avoid the code duplication?

axxel commented on core/src/qrcode/QRFormatInformation.h:
```diff
@@ -24,19 +24,21 @@ namespace ZXing {
 public:
 	FormatInformation() = default;
 
-	static FormatInformation DecodeFormatInformation(uint32_t formatInfoBits1, uint32_t formatInfoBits2);
```
I fail to understand why this "Factory" abstraction was introduced. It's late, so maybe that is the reason... ;) But why not simply add a `DecodeFormatInformation(uint32_t formatInfoBits)` overload that produces the MircoQRCode info?

axxel commented on test/blackbox/BlackboxTestRunner.cpp:
```diff
@@ -658,6 +658,13 @@ int runBlackBoxTests(const fs::path& testPathPrefix, const std::set<std::string>
 			{ 0, 0, pure },
 		});
+
+		runTests("microqrcode-1", "MicroQRCode", 13, {
```
All the binary image files need to be compressed and potentially cropped heavily. The test image directory already takes up way to much space and people are rightly complaining about that. I can take care of that with the tools I used before and provide the results by committing to your PR branch. Never tried that but I believe that should work somehow ;). Could also be useful for other 'merging' work discussed above.

corbers replied:
Could you write up the process and the tools you use?  I have more images to add which might test different features of the micro QR code and so it would be good if I could process the images before committing.

axxel replied:
I was experimenting with different online jpg minimization services. I'll happily do the cropping and manual tuning. I'd in general try to keep the number of new test images as low as reasonable.

Background: to 'fix' this issue I was also exploring the use of JpegXL to recode all the samples and thereby maybe reduce their size to ideally something like a 10th. But that was not important enough yet and replacing the pngs with jpegs only to maybe replace them again later only further bloats the git-repo size itself. (starting with png encoded Jpegs was just not a good idea by the Java people in the first place, in my opinion ;)).

axxel reviewed (changes requested):
It got late, so sorry if I should have written stuff that looks like lacking attention to detail ;). I definitively see opportunity to further reduce code duplication and am convinced that 'merging' those two variants of more or less the same code, is the right way to go.

corbers commented on core/CMakeLists.txt:
```diff
@@ -206,6 +206,43 @@ if (BUILD_READERS)
+)
+if (BUILD_READERS)
+    set (MICROQRCODE_FILES ${MICROQRCODE_FILES}
+        src/qrcode/MQRBitMatrixParser.h
```
Sorry about this.  Started using a text editor instead of my IDE and so tabs were inserted.

corbers commented on core/src/BarcodeFormat.h:
```diff
@@ -43,13 +43,14 @@ enum class BarcodeFormat
-	QRCode          = (1 << 13), ///< QR Code (2D)
-	UPCA            = (1 << 14), ///< UPC-A (1D)
-	UPCE            = (1 << 15), ///< UPC-E (1D)
+	MicroQRCode     = (1 << 12), ///< Micro QR Code (2D)
```
Done

corbers commented on core/src/BitSource.cpp:
```diff
@@ -1,19 +1,19 @@
-* See the License for the specific language governing permissions and
-* limitations under the License.
-*/
+ * Copyright 2016 Nu-book Inc.
```
OK.  I've removed the formatting changes.

corbers commented on core/src/BitSource.cpp:
```diff
@@ -24,54 +24,68 @@
 
-int
-BitSource::available() const
+static int ReadBitsImpl(int numBits, const ByteArray& bytes, int& byteOffset, int& bitOffset)
```
ReadBitsImpl is a new function, not a member function belonging to the BitSource class.  I've checked the rest of the project and leading underscores are not used for function parameters so I think what I've done is correct.

corbers replied:
The reason `peakBits` is required is that the terminator code is longer than the mode codes. In QR codes the mode codes and the terminator code are all 4 bits.  This is great because we can read 4 bits from the data stream to see if the mode has changed or we should terminate.  In micro QR codes the mode codes vary in length but the terminator code is always longer than the mode code.
- M1 - 0 length mode code, 3 bits terminator code
- M2 - 1 bit mode code, 5 bits terminator code
- M3 - 2 bit mode code, 7 bits terminator code
- M4 - 3 bit mode code, 9 bits terminator code

The numeric mode code is 1, 2 or 3 0s and the terminator code is 3, 5, 7, or 9 0s.  When reading the bit stream we need to peak ahead to see if we're about to parse a terminator code or just a numeric code.  If the peak is not a terminator code then it will read the bits as normal.

axxel replied:
Thanks. Very informative. Could you maybe put this info into a comment at an appropriate place, now that it is readily available?

corbers commented on core/src/qrcode/MQRBitMatrixParser.cpp:
```diff
@@ -0,0 +1,121 @@
+	// in a Version M3-M symbol is a 2x2 square 4-module block.
+	// See ISO 18004:2006 6.7.3.
+	bool hasD4mBlock = version.versionNumber() % 2 == 1;
+	int d4mBlockIndex = version.versionNumber() == 1                                                    ? 3
```
Agreed.  Your updated clang-format file improves things.

corbers commented on core/src/qrcode/MQRCodecMode.h:
```diff
@@ -0,0 +1,66 @@
+class Version;
+
+/**
+ * <p>See ISO 18004:2006, 6.4.1, Tables 2 and 3. This enum encapsulates the various modes in which
```
These comments are correct.  Tables 2 and 3 list both micro and QR mode indicators and number of bits in character count.  I checked before committing.

axxel replied:
Thanks and sorry for implying otherwise.

corbers commented on core/src/qrcode/MQRDataMask.h:
```diff
@@ -0,0 +1,52 @@
+namespace MicroQRCode {
+
+/**
+ * <p>Encapsulates data masks for the data bits in a micro QR code, per ISO 18004:2006 6.8.</p>
```
No.  References and comments are correct.

corbers commented on core/src/qrcode/MQRDataBlock.cpp:
```diff
@@ -0,0 +1,89 @@
+
+	// First count the total number of data blocks
+	int totalBlocks = ecBlocks.numBlocks();
+	if (totalBlocks == 0)
+		return {};
```
The reason for the duplication is because there are two Version classes - `QRVersion` and `MQRVersion` - which means that there needs to be two `GetDataBlock` functions.  What I'll do is change the construction functions `ProvisionalVersionForDimension,` `VersionForNumber,` `DecodeVersionInformation` to have a bool flag to indicate if a micro QR code is being processed.  If true, I can return a micro QR code version of QRVersion for use in the rest of the decoding process.  This should simplify the rest of the decoding code.

🤞 my IDE doesn't try to reformat all the code using clang-format as I make these changes 😄

axxel replied:
Perfect. That sounds like what I had in mind in my comment from yesterday. About the clang-format issue: removing the file while working on this might do the trick? Or what I use extensively during commit is `git gui` and its ability to add individual hunks or even lines to the staging area. Thanks for the effort!

axxel commented:
There is still whitespace noise in the diffs but part of that in files you just announced to change again anyway (but still also in the CMakeLists.txt). In terms of merging the code and preventing code duplication, this looks like the right direction. One area looks a bit inconsistent right now is that parts of the differences are handled by the boolean flag and other as equally named functions from different namespaces. Once your planned changes are done, we should look at it again and see what can still be done to make the code maybe even more clean/consistent.

axxel commented:
If you'd allow me to push commits to your PR branch (see [here](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests)), I could replace the images and provide a few minor improvements that I can type faster myself than I can communicate them to you.

corbers commented:
> If you'd allow me to push commits to your PR branch (see [here](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests)), I could replace the images and provide a few minor improvements that I can type faster myself than I can communicate them to you.

Unfortunately this doesn't work because the fork is in a repository owned by an organisation and not an individual.  There doesn't appear to be a workaround.  See https://github.community/t/how-can-we-enable-allow-edits-from-maintainers-by-default/2847

corbers commented:
I've checked with my company and we can give you write access to our fork of zxing.  That should speed up the integration of the micro QR code support.  Is that OK with you?

axxel commented:
Sure. If you need need to get in touch with me 'in private' for that, just send me an e-mail (to be found in the git logs).

axxel commented:
As you can see, I've worked on further reducing the diff between this PR and master. It is now easier to see what substantially changed. Next would be the work on the FinderPatternFinder and friends code. I do now understand a bit better what algorithmic ideas were used and also made it run on my phone which let me very interactively test the limitations regarding rotation and perspective transformation of the symbol. The results were consistent which what I expected earlier when I saw that the `DetectWhiteRect` function was used. I do believe that the situation could be substantially improved by incorporating another approach altogether. If that works out, my gut feeling right now is that not much of the current detector code will survive. So I don't think it makes sense to work on that right now (like the earlier mentioned `ResultPoint -> Point<T>` conversion).
Since I reimplemented much of the current QRCode detector, I'm probably in a better position to hack something together based on that infrastructure. So I suggest I give it a go and we'll see whether I can deliver on my gut-feeling based promise. Thoughts? Objections?

Note: the build failure is apparently a githup actions issue. The current master HEAD fails to build as well, even though the same code worked just fine 2 days ago.

corbers commented:
The code is looking better.  Thanks for your help.
I've no objections to you having a go at improving the detector - that's one of the reasons why we shared our code on GitHub.  It's a bit harder than the QR code in that there's only a single finder pattern to help with the calculation of the transform for the rotation and perspective correction.  I hope your ideas work out.
I'm going to add in a comment about why we need to check for the terminator code and peak at the bits.  The decoder is also missing the ability to read the version number from the micro QR code.  This is part of the standard and it's something we've missed.  I'll try to add this in today.

axxel commented on core/src/qrcode/QRBitMatrixParser.cpp:
```diff
@@ -50,8 +50,21 @@ const Version* ReadVersion(const BitMatrix& bitMatrix, bool isMicro)
+			return theParsedVersion;
+		return nullptr;
+	}
+
```
Forgive me if I'm wrong but I don't see how this change improves the detection rate in any way. It actually decreases it: you simply check the integrity of the two timing patterns and fail the decoding if just any one of those bits is flipped. This basically reduces the error tolerance of the symbol to 0% if the error occurs inside the timing pattern. What was your intention for this change, if I may ask?

The version of a MicroQRCode is in fact encoded inside the format info and that is where I intend to parse it from in my new detector implementation. (This whole concept of strictly separating detection and decoding of the symbol is ill advised here. See also the suggested 'reference algorithm' in the specification.)

axxel replied:
I see what you mean. It still does not make sense to me, neither does the reference algo, to be fair. They are not doing the same thing though: The spec algo differs in that it does only enforce the number of dark modules of both timing patterns to be equal and between 2 and 5 (only to ignore that info later on anyway). Your implementation requires them to be both exactly as anticipated by the version number estimate determined by other means. Anyway, sorry if this sounds like dogmatism.

axxel replied:
Update: now that I tried your code for the first time after I've got my first results from my new detector, I noticed that it is plain wrong: you sampled the wrong bits and use the masks for QR instead of MQR codes. So this code will indeed go away. ;).

corbers commented:
I’m happy to follow your lead on this. I have limited experience of
processing real world codes and it might be that my approach is too strict.
I still can’t believe that masked and unmasked bits are used to determine
the format information. This effectively increases the hamming distance
threshold from 3 to something much higher. I can’t find any information on
why this is required in either the C++ or Java ZXing projects other than
the code comment that sometimes the bits don’t have their mask. I guess
this is another example of a real world compromise between standard
adherence and usability. 😀


On Fri, 13 May 2022 at 18:14, axxel ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In core/src/qrcode/QRBitMatrixParser.cpp
> <https://github.com/nu-book/zxing-cpp/pull/324#discussion_r872614873>:
>
> > @@ -50,8 +50,21 @@ const Version* ReadVersion(const BitMatrix& bitMatrix, bool isMicro)
>  	int dimension = bitMatrix.height();
>
>  	int provisionalVersion = (dimension - Version::DimensionOffset(isMicro)) / Version::DimensionStep(isMicro);
> +	if (isMicro) {
> +		int verticalVersionBits = 0;
> +		for (int y = 8; y < dimension; ++y)
> +			AppendBit(verticalVersionBits, getBit(bitMatrix, 0, y, false));
> +		int horizontalVersionBits = 0;
> +		for (int x = 8; x < dimension; ++x)
> +			AppendBit(horizontalVersionBits, getBit(bitMatrix, x, 0, false));
> +		auto theParsedVersion = Version::DecodeVersionInformation(verticalVersionBits, horizontalVersionBits);
> +		if (theParsedVersion != nullptr && theParsedVersion->dimensionForVersion() == dimension)
> +			return theParsedVersion;
> +		return nullptr;
> +	}
> +
>
> I see what you mean. It still does not make sense to me, neither does the
> reference algo, to be fair. They are not doing the same thing though: The
> spec algo differs in that it does only enforce the number of dark modules
> of both timing patterns to be equal and between 2 and 5 (only to ignore
> that info later on anyway). Your implementation requires them to be both
> exactly as anticipated by the version number estimate determined by other
> means. Anyway, sorry if this sounds like dogmatism.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/nu-book/zxing-cpp/pull/324#discussion_r872614873>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHK6U7QIUWGGMQWQGGDICJTVJ2EYLANCNFSM5UPLFSZA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

corbers commented:
Like I said I’m happy for the code to go but I’m not sure how it’s “plain
wrong”. Are you referring to the read version changes I introduced at the
end of last week?  If so, maybe you can explain how I’m not sampling the
timing pattern along the edge of the code to determine the version number?

On Sun, 15 May 2022 at 16:44, axxel ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In core/src/qrcode/QRBitMatrixParser.cpp
> <https://github.com/nu-book/zxing-cpp/pull/324#discussion_r873187816>:
>
> > @@ -50,8 +50,21 @@ const Version* ReadVersion(const BitMatrix& bitMatrix, bool isMicro)
>  	int dimension = bitMatrix.height();
>
>  	int provisionalVersion = (dimension - Version::DimensionOffset(isMicro)) / Version::DimensionStep(isMicro);
> +	if (isMicro) {
> +		int verticalVersionBits = 0;
> +		for (int y = 8; y < dimension; ++y)
> +			AppendBit(verticalVersionBits, getBit(bitMatrix, 0, y, false));
> +		int horizontalVersionBits = 0;
> +		for (int x = 8; x < dimension; ++x)
> +			AppendBit(horizontalVersionBits, getBit(bitMatrix, x, 0, false));
> +		auto theParsedVersion = Version::DecodeVersionInformation(verticalVersionBits, horizontalVersionBits);
> +		if (theParsedVersion != nullptr && theParsedVersion->dimensionForVersion() == dimension)
> +			return theParsedVersion;
> +		return nullptr;
> +	}
> +
>
> Update: now that I tried your code for the first time after I've got my
> first results from my new detector, I noticed that it is plain wrong: you
> sampled the wrong bits and use the masks for QR instead of MQR codes. So
> this code will indeed go away. ;).
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/nu-book/zxing-cpp/pull/324#discussion_r873187816>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHK6U7UQOJ3RIC6Z5TI2EDLVKELVLANCNFSM5UPLFSZA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

axxel commented:
It seems you did not (yet) notice my "Update²: What I just posted was complete BS. Sorry for the noise...". So the only thing that was "plain wrong" was my comment. (How embarrassing! I could explain how I came this wrong conclusion but it is really not worth the words and neither of our time...).

To address your comment about the masking (or lack thereof) of the format info bits: That is indeed just bloody annoying. I had a little bit of hope that the MicroQRCode variant did not require this hack (somehow thinking that it is likely newer and encoder implementations therefore likely less buggy). But the very first sample I was working with seems to have exactly the same problem. And yes: this reduces the effectiveness of the error detection/recovery by a whole bit (if I'm not mistaken). Especially annoying as it currently seems to ruin my idea of how to determine the symbol version and orientation by sampling the format info bits during detection. Still a work in progress.

corbers commented:
Thanks for the update and clarification.  I was about to re-assess my career in coding 😆 

I don't think there is anything I can contribute at the moment but if I can help in any way then let me know.  I'm happy to test, code review etc.  Anything to help get this PR finished.

corbers commented:
I've run the new code on all the micro QR code samples we have for automated testing.  There are 21 in total of which I think 6 are currently committed in this project.
- 3 of the 21 are pure barcodes and decode with no problems.
- In the remaining 18 there are 4 failures at 0 degrees rotation.  3 of these failures match what was expected with the old code.  The remaining failure, 12.png, decodes using the old code but fails with the new code.  However, the cropped and compressed version of this same image found in the project repository does decode correctly.
- I have a second set of this data where the first 11 images have been rotated 90 degrees clockwise.  The results from these match the results from the new code.
- The third set contains codes rotated by 45 degrees.  Again there are 11 of these.  With the old code all 11 failed but with the new code 6 fail at fast speed and 4 at slow.

My conclusion is that these code changes have either kept the performance the same or improved things.  I think the new code changes have made the implementation simpler to understand and should make it easier to support in the future.
(I can provide copies of the images that caused problems in both code versions if you wish).

Lastly, Visual Studio complains about this code from ConcentricFinder.cpp
`std::array lines{RegressionLine{corners[0] + 1, corners[1]}, RegressionLine{corners[1] + 1, corners[2]}, RegressionLine{corners[2] + 1, corners[3]}, RegressionLine{corners[3] + 1, &(*points.end())}};` because `points.end()` points to an element beyond the end of the vector.  For the tests I ran I used `points.end() - 1` instead.

axxel commented:
Thanks for the testing. I would be interested in getting my hands on the 4 failures with 0° rotation and the 4 failures with 45° rotation. The issue with 12.jpg is basically the same as with the one extra image I added that currently does not process (put there as a reminder for future enhancements). Can you put those 8 images somewhere online to download or simply send them via e-mail?

axxel commented:
I got it and will have a look at what I might be able to improve further. Thank you for your cooperation and help, too! It was a pleasure.
